trusts/decorators.py

```diff
@@ -119,4 +194,31 @@ def check_perms(request, *args, **kwargs):
             raise PermissionDenied
         return False
 
+    def check_by_p(request, *args, **kwargs):
+        perms = perm.get_leaves()
+
+        for p in perms:
+            fieldlookups = _resolve_fieldlookups(request, kwargs, **p.get_lookup_kwargs())
+            items = None
+            if fieldlookups is not None:
+                items = _get_permissible_items(request, p.perm, fieldlookups)
+                if items is None:
+                    p.permitted = False
```

Comment:
When you use "P('app.moderate_xyz', pk=K('xyz_id')) | P('app.update_xyz:own', pk=K('xyz_id'))" you create a new P instance every time, so it doesn't matter will we modify it or not. Django can't cache this, because it has to call view_function every time (when url matches of course)